File: 2019/elixir/01.exs
#! /usr/bin/env elixir

defmodule Day01 do
  def fuel(mass), do: Integer.floor_div(mass, 3) - 2
  def recursive_fuel(mass, acc\\0) do
    case fuel(mass) do
        x when x <= 0 -> acc
        next -> recursive_fuel(next, acc+next)
    end
  end
end


masses = File.stream!("../input/01")
  |> Stream.map(&String.trim_trailing/1)
  |> Enum.map(&String.to_integer/1)

masses 
  |> Enum.map(&Day01.fuel/1) 
  |> Enum.sum 
  |> IO.inspect(label: "Part 1")
masses 
  |> Enum.map(&Day01.recursive_fuel/1) 
  |> Enum.sum 
  |> IO.inspect(label: "Part 2")

File: 2019/python/01.py
import math

def part1(masses):
    return sum(fuel_for_mass(x) for x in masses)

def part2(masses):
    return sum(recursive_fuel_for_mass(x) for x in masses)

def fuel_for_mass(mass):
    return math.floor(mass/3) - 2    

def recursive_fuel_for_mass(mass, acc=0):
    next = fuel_for_mass(mass)
    return acc if next <= 0 else recursive_fuel_for_mass(next, acc+next)


if __name__ == "__main__":
    with open("../input/01") as f:
        masses = [int(x) for x in f.readlines()]
        print('Part 1: {}'.format(part1(masses)))
        print('Part 2: {}'.format(part2(masses)))